File: tests/java_static/src/test/java/myairtable/TestClientUrlEncoding.java
package myairtable;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.net.URI;
import java.util.List;
import org.junit.jupiter.api.Test;

/**
 * J-port of Kotlin K2.3 verify-first spike, kept as a regression test: Airtable's server decodes a
 * raw {@code +} in query values as a space, corrupting formulas like {@code LEN({f})+1}. The Java
 * client uses {@link java.net.URLEncoder} form encoding, which percent-encodes {@code +} as {@code
 * %2B} (and encodes spaces as {@code +}, which Airtable decodes back to spaces) — these tests pin
 * that behavior. Offline: only builds URLs, no requests are sent.
 */
class TestClientUrlEncoding {

  private final AirtableClient client = new AirtableClient("appBASE", "key");

  @Test
  void plusInFormulaIsPercentEncoded() {
    URI url =
        client.tableUrl(
            "tblX",
            new AirtableQuery()
                .withFormula("LEN({f})+1 = 2")
                .withReturnFieldsByFieldId(false)
                .toParameters());
    String query = url.getRawQuery();
    assertTrue(query.contains("%2B"), "raw + must be percent-encoded, got: " + query);
    // No RAW plus may remain except as the space encoding.
    assertFalse(query.contains("})+1"), "raw + leaked into query: " + query);
  }

  @Test
  void urlPathContainsBaseAndTable() {
    URI url = client.tableUrl("tblX", List.of());
    assertEquals("/v0/appBASE/tblX", url.getRawPath());
  }

  @Test
  void recordUrlAppendsRecordId() {
    URI url = client.recordUrl("tblX", "recABC", List.of());
    assertEquals("/v0/appBASE/tblX/recABC", url.getRawPath());
  }

  @Test
  void deleteRecordsParamsEncodeAsRepeatedRecordsArray() {
    URI url =
        client.tableUrl(
            "tblX",
            List.of(
                new AirtableQuery.Param("records[]", "rec1"),
                new AirtableQuery.Param("records[]", "rec2")));
    String query = url.getRawQuery();
    assertTrue(query.contains("records%5B%5D=rec1") || query.contains("records[]=rec1"), query);
    assertTrue(query.contains("rec2"));
  }
}

File: tests/kotlin_static/src/test/kotlin/myairtable/TestClientUrlEncoding.kt
package myairtable

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

/**
 * K2.3 verify-first spike, kept as a regression test: Airtable's server
 * decodes a raw `+` in query values as a space, corrupting formulas like
 * `LEN({f})+1`. The Swift target needed a manual `%2B` force-encode; Ktor's
 * URLBuilder already percent-encodes `+` (and encodes spaces as `+`, which
 * Airtable decodes back to spaces) — these tests pin that behavior so a Ktor
 * upgrade can't silently regress it.
 */
class TestClientUrlEncoding {
    private val client = AirtableClient(baseId = "appBASE", apiKey = "key")

    @Test
    fun plusInFormulaIsPercentEncoded() {
        val url = client.tableUrl("tblX", AirtableQuery(formula = "LEN({f})+1 = 2", returnFieldsByFieldId = false).toParameters())
        val query = url.encodedQuery
        assertTrue("%2B" in query, "raw + must be percent-encoded, got: $query")
        // No RAW plus may remain except as the space encoding.
        assertFalse("})+1" in query, "raw + leaked into query: $query")
    }

    @Test
    fun urlPathContainsBaseAndTable() {
        val url = client.tableUrl("tblX")
        assertEquals("/v0/appBASE/tblX", url.encodedPath)
    }

    @Test
    fun recordUrlAppendsRecordId() {
        val url = client.recordUrl("tblX", "recABC")
        assertEquals("/v0/appBASE/tblX/recABC", url.encodedPath)
    }

    @Test
    fun deleteRecordsParamsEncodeAsRepeatedRecordsArray() {
        val url = client.tableUrl("tblX", listOf("records[]" to "rec1", "records[]" to "rec2"))
        val query = url.encodedQuery
        assertTrue("records%5B%5D=rec1" in query || "records[]=rec1" in query, query)
        assertTrue("rec2" in query)
    }
}
